tutorials/git-advanced.rst

```diff
@@ -0,0 +1,604 @@
+       printf("Implemented task 1\n");
+   }
+
+And your partner’s code might look like this:
+
+::
+
+   int foo()
+   {
+       printf("Implemented task 2\n");
+   }
+
+This actually represents a *conflict* in your code: you each have
+divergent replacements for the ``// Write your code here`` line, and it
+is not clear which version is the correct one. In fact, while there may
+be cases where we simply want to use one version over another, in this
+case we would like to *merge* these two versions together to produce
+something like this:
+
+::
+
+   int foo()
+   {
+       printf("Implemented task 1\n");
+       printf("Implemented task 2\n");
+   }
+
+Notice how, earlier, there was no conflict when implementing ``bar()``
+and ``baz()`` because you were adding code to ``hw1.c``, instead of
+replacing existing code with different versions.
+
+Version control systems like Git are very useful when dealing with
+scenarios like the ones above. They will allow two (or more) developers
+to work concurrently on the same code and, whenever possible, will
+automatically merge together changes that do not conflict. When a
+conflict does arise, Git provides a specific mechanism to resolve that
+conflict, which we discuss in the following sections.
+
+Branches
+--------
+
+So far, the commits in your tutorial repository have created a linear
+sequence of changes like this:
+
+.. figure:: git-branches.png
+   :alt: branches
+
+This is known as the ``main`` branch of your repository (Git itself uses
+the name ``master`` by default, but GitHub `uses the name "main" <https://github.com/github/renaming>`__).
+But, what
+is a "branch"? A branch, loosely, is an
+**independent commit history** than can be manipulated in its own right.
+So far, you have been working with only one branch (the ``main`` branch) and,
+thus, with a single linear history of commits.
+
+However, Git (and most version control systems) allow you to create
+*multiple* branches. These branches are not completely independent of
+each other as, ordinarily, a branch must be “branched off” from an
+existing commit. So, for example, we could have the following:
+
+.. figure:: git-branches-2.png
+   :alt: branches2
+
+Here, besides the ``main`` branch, we have an additional ``foo`` branch
+which could be used to work on a separate feature of a project (or, in this
+case, on an additional ``foobar.c`` file where we are implementing a ``foo()``
+function).  This
+separate branch allows us to work on this task independently from other
+tasks; this may seem over-complicated, but suppose you were working on
+a homework or project with a classmate: branches would allow you to work independently
+without having to step on each other’s toes. This is similar to the
+first example we showed above (where two developers could be working on
+separate functions, ``bar()`` and ``baz()``).
+
+To experiment with branches, start by adding a file called ``echo.c``
+to your repository, with the following contents:
```

Comment:
You should clarify if this is the repo from the first half of the tutorial.

Reply:
I've added a note in the TODO at the top saying that we'll need to tell them to create an empty repository for this tutorial (if they haven't worked through the Git Basics tutorial)